xdsl/dialects/dlti.py

```diff
@@ -47,12 +71,39 @@ class DLTIEntryMap(ParametrizedAttribute, ABC):
     # In MLIR, this is a DataLayoutEntryInterface.
     entries: ArrayAttr[DataLayoutEntryAttr]
 
+    def __init__(
+        self,
+        contents: ArrayAttr
+        | list[DataLayoutEntryAttr]
+        | dict[StringAttr | TypeAttribute | str, Attribute | str | int | float],
+    ):
+        if isinstance(contents, dict):
+            contents = ArrayAttr(
+                [DataLayoutEntryAttr(k, v) for k, v in contents.items()]
+            )
+        elif isinstance(contents, list):
+            contents = ArrayAttr(contents)
+
+        assert isinstance(contents, ArrayAttr)
+
+        super().__init__(contents)
```

Comment:
I would recommend making a few changes:
1. only allowing two posible types, the actual attribute to just assign or a pretty python type
2. Only allowing immutable types as inputs as these will be covariant, and will avoid pyright headaches down the road
```suggestion
        contents: ArrayAttr[DataLayoutAttr]
        | Mapping[StringAttr | TypeAttribute | str, StringAttr | IntegerAttr | FloatAttr[Float32Type] | str | int | float],
    ):
        if not isa(contents, ArrayAttr[DataLayoutAttr]):
            contents = ArrayAttr(
                [DataLayoutEntryAttr(k, v) for k, v in contents.items()]
            )

        super().__init__(contents)
```

This requires `from typing import Mapping` at the top of the file, it's the immutable protocol version of `dict`.

Reply:
I've moved to using Mapping as suggested, and have removed list[DataLayoutEntryAttr], now have  ArrayAttr[DataLayoutAttr] and the mapping as suggested here.